### DISCUSSION TITLE
Borderless window not focusing properly

### DISCUSSION THREAD

**Discussion by aekobear** [fltk-rs/fltk-rs#1033] [Q&A] [answered]
Hi! I'm trying to create a simple tray app with a single FLTK window which acts as a pop-up menu and closes as soon as it loses focus

(MS Windows x64, fltk-rs 1.2.19)

The following code works perfectly until I add `set_border(false)`. When the border is removed, the window says it has focus (IE the `Focused` event fires), but it doesn't actually have focus and I can't get the `Unfocus` event to fire without first clicking on the window.

```rust
use fltk::{app, prelude::*, window::Window, enums::{Color, Event}};

fn main() {
    let app = app::App::default();
    let mut wind = Window::new(100, 100, 200, 100, "FLTK Test");
    wind.set_border(false); // <-- this is the problem
    wind.handle(move |widget, ev| -> bool {
        match ev {
            Event::Focus => { println!("FOCUSED"); }
            Event::Unfocus => { println!("BLURRED"); }
            e => { }
        }
        true
    });
    wind.end();
    wind.show();
    wind.wait_for_expose();
    app.run().unwrap();
}
```

I wasn't sure whether this is a bug in the rust implementation, a bug with FLTK itself, or a sign that maybe I'm doing something wrong and there's a better way to close an application when the user clicks outside the window.

**Comment by MoAlyousef** [comment, marked as answer]
A small repro in C++:
```c++
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Enumerations.H>
#include <iostream>

struct MyWindow : Fl_Window {
  MyWindow(int x, int y, int w, int h, const char *l = 0): Fl_Window(x, y, w, h, l) {}
  virtual int handle(int e) override {
    switch (e) {
      case FL_FOCUS: {
        std::cout << "Focused" << std::endl;
        return 1;
      }
      case FL_UNFOCUS: {
        std::cout << "Unfocused" << std::endl;
        return 1;
      }
      default: return Fl_Window::handle(e);
    }
  }
};

int main() {
  auto win = new MyWindow(100, 100, 400, 300, "Window");
  win->border(false);
  win->end();
  win->show();
  return Fl::run();
}
```
I think it’s none of the above. I think it’s a windows behavior since borderless windows aren’t considered the application window by default, unless you pass the win32 WS_EX_APPWINDOW flag to GWL_EXSTYLE. 

you should probably ask in the fltk general mailing list:
https://groups.google.com/g/fltkgeneral

**Reply by aekobear** [reply]
Thanks, you were absolutely right! I bit the bullet and messed with the winapi for a while and was able to get the desired behavior by setting  `GWL_STYLE` to `WS_POPUP | WS_VISIBLE` (gives me a borderless window with focus), as well as setting `GWL_EXSTYLE` to `WS_EX_TOOLWINDOW` (prevents it from appearing in the taskbar)

**Comment by MadameMinty** [comment]
While this is an old issue, and I'd prefer to avoid necromancy, I've found this via google as one of the top results, but it doesn't help.

I've encountered the exact same problem using `win.set_shape()`, which does the equivalent of `win.set_border(false);` under the hood before setting the window shape. This removes initial focus from the window, and ruins Unfocus detection.

I've applied the suggested resolution i.e. hoisting the flags:

```rust
use winapi::um::winuser::{
    GetWindowLongPtrW, SetWindowLongPtrW,
    GWL_STYLE, WS_POPUP, WS_VISIBLE,
    GWL_EXSTYLE, WS_EX_TOOLWINDOW};
use winapi::shared::windef::HWND;

...

    unsafe {
        let hwnd = win.raw_handle() as HWND;
        // borderless window with focus
        let style = GetWindowLongPtrW(hwnd, GWL_STYLE);
        SetWindowLongPtrW(hwnd, GWL_STYLE,
            (style | WS_POPUP as isize | WS_VISIBLE as isize) as isize);
        // prevent from appearing in the taskbar
        let exstyle = GetWindowLongPtrW(hwnd, GWL_EXSTYLE);
        SetWindowLongPtrW(hwnd, GWL_EXSTYLE,
            (exstyle | WS_EX_TOOLWINDOW as isize) as isize);
    }
```

But while WS_EX_TOOLWINDOW removed the window from the taskbar, WS_POPUP and WS_VISIBLE didn't seem to remove the border, nor bring focus. Unsure why.

What did work, however, was [this neat thing](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setactivewindow):

```rust
use winapi::um::winuser::SetActiveWindow;
use winapi::shared::windef::HWND;

...

    let hwnd = win.raw_handle() as HWND;
    unsafe { SetActiveWindow(hwnd); }
```

`SetForegroundWindow` [also works](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setforegroundwindow). Either way, there seems to be no avoiding the use of `winapi`.

**Reply by MoAlyousef** [reply]
Thank you for sharing this solution. It does work better :)
